 
 <launch>
  <!-- Extended Kalman filter fusion - odometry -->
    <node pkg="robot_localization" type="ekf_localization_node" name="ekf_rel" clear_params="true" >
        <param name="frequency" value="50"/>
        <param name="sensor_timeout" value="0.03"/>
        <param name="two_d_mode" value="true"/>
      
        <param name="map_frame" value="map"/>
        <param name="odom_frame" value="mechROS/tf/odom"/>
        <param name="base_link_frame" value="mechROS_base_link"/>
        <param name="world_frame" value="mechROS/tf/odom"/>
      
        <param name="odom0" value="/mechROS/odom"/>
        <!-- <param name="imu0" value="/imu/data_corrected"/> -->
        <param name="imu0" value="/imu/data"/>

        <!-- Use control to predict acceleration -->
        <param name="use_control" value="true"/>
        <param name="stamped_control" value="false"/>
        <param name="control_timeout" value="0.02"/>
        <rosparam param="control_config">[true,  false, false,
                                  false, false, true]</rosparam>
        <rosparam param="acceleration_limits">[1.0, 0.0, 0.0,
                                       0.0, 0.0, 7.5]</rosparam>
        <rosparam param="decceleration_limits">[1.0, 0.0, 0.0,
                                       0.0, 0.0, 7.5]</rosparam>
        
        <rosparam param="odom0_config">[false,  false,  false,
                                        false,  false,  false,
                                        true,  true,  false,
                                        false,  false,  true,
                                        false,  false,  false]</rosparam>

        <rosparam param="imu0_config">[false,  false,  false,
                                        false,  false,  false,
                                        false,  false,  false,
                                        false,  false,  true,
                                        false,  false,  false]</rosparam>
        
        
        <param name="odom0_differential" value="false"/>
        <param name="imu0_differential" value="false"/>
        
        <param name="odom0_relative" value="false"/>
        <!-- <param name="imu0_relative" value="true"/> -->
        
        <param name="odom0_queue_size" value="10"/>
        <param name="imu0_queue_size" value="10"/>

        <!-- Thresholdy - otestovat -->
        <!-- <param name="odomN_twist_rejection_threshold" value="10"/> -->
        <!-- <param name="imuN_angular_velocity_rejection_threshold" value="10"/> -->
        <!-- <param name="imuN_linear_acceleration_rejection_threshold" value="10"/> -->

      
        <param name="debug"           value="true"/>
        <param name="debug_out_file"  value="/home/patrik/catkin_ws/src/mech_ros/debug/debug_ekf_rel.txt"/>
      
        <rosparam param="process_noise_covariance">[0.05, 0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                    0,    0.05, 0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                    0,    0,    0.06, 0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                    0,    0,    0,    0.05, 0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                    0,    0,    0,    0,    0.05, 0,    0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                    0,    0,    0,    0,    0,    0.01, 0,     0,     0,    0,    0,    0,    0,    0,    0,
                                                    0,    0,    0,    0,    0,    0,    0.05, 0,     0,    0,    0,    0,    0,    0,    0,
                                                    0,    0,    0,    0,    0,    0,    0,     0.05, 0,    0,    0,    0,    0,    0,    0,
                                                    0,    0,    0,    0,    0,    0,    0,     0,     0.04, 0,    0,    0,    0,    0,    0,
                                                    0,    0,    0,    0,    0,    0,    0,     0,     0,    0.01, 0,    0,    0,    0,    0,
                                                    0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0.01, 0,    0,    0,    0,
                                                    0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0.09, 0,    0,    0,
                                                    0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0.012, 0,    0,
                                                    0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0.01, 0,
                                                    0,    0,    0,    0,    0,    0,    0,     0,     0,    0,    0,    0,    0,    0,    0.015]</rosparam>
      
        <param name="dynamic_process_noise_covariance" value="true"/>

        <rosparam param="initial_estimate_covariance">[1e-9, 0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                       0,    1e-9, 0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                       0,    0,    1e-9, 0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                       0,    0,    0,    1e-9, 0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                       0,    0,    0,    0,    1e-9, 0,    0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                       0,    0,    0,    0,    0,    1e-9, 0,    0,    0,    0,     0,     0,     0,    0,    0,
                                                       0,    0,    0,    0,    0,    0,    1e-9, 0,    0,    0,     0,     0,     0,    0,    0,
                                                       0,    0,    0,    0,    0,    0,    0,    1e-9, 0,    0,     0,     0,     0,    0,    0,
                                                       0,    0,    0,    0,    0,    0,    0,    0,    1e-9, 0,     0,     0,     0,    0,    0,
                                                       0,    0,    0,    0,    0,    0,    0,    0,    0,    1e-9,  0,     0,     0,    0,    0,
                                                       0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     1e-9,  0,     0,    0,    0,
                                                       0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     1e-9,  0,    0,    0,
                                                       0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     1e-9, 0,    0,
                                                       0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    1e-9, 0,
                                                       0,    0,    0,    0,    0,    0,    0,    0,    0,    0,     0,     0,     0,    0,    1e-9]</rosparam>
        
            <!-- Initial estimation - delete after testing -->
        <rosparam param="initial_state">[ 0.0,  0.0,  0.0,
                                          0.0,  0.0,  0.0,
                                          0.0,  0.0,  0.0,
                                          0.0,  0.0,  0.0,
                                          0.0,  0.0,  0.0]</rosparam>
    
        <param name="publish_tf" value="true"/>
        <remap from="/odometry/filtered" to="/odometry_rel/filtered" />
        <param name="publish_acceleration" value="true"/>
        <remap from="/accel/filtered" to="/accel_rel/filtered" />
    </node>
</launch>